smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/PaginatorGenerator.kt

```diff
@@ -0,0 +1,245 @@
+            writer,
+            serviceSymbol,
+            paginatedOperation,
+            inputSymbol,
+            outputSymbol,
+            paginationInfo,
+            cursorSymbol
+        )
+
+        // Optionally generate paginator when nested item is specified on the trait.
+        if (itemDesc != null) {
+            renderItemPaginator(
+                writer,
+                service,
+                paginatedOperation,
+                itemDesc,
+                outputSymbol
+            )
+        }
+    }
+
+    // Generate the paginator that iterates over responses
+    private fun renderResponsePaginator(
+        writer: KotlinWriter,
+        serviceSymbol: Symbol,
+        operationShape: OperationShape,
+        inputSymbol: Symbol,
+        outputSymbol: Symbol,
+        paginationInfo: PaginationInfo,
+        cursorSymbol: Symbol
+    ) {
+        val nextMarkerLiteral = paginationInfo.outputTokenMemberPath.joinToString(separator = "?.") {
+            it.defaultName()
+        }
+        val markerLiteral = paginationInfo.inputTokenMember.defaultName()
+
+        writer.write("")
+        writer.dokka(
+            """
+                  Paginate over [${outputSymbol.name}] results.
+                  When this operation is called, a [kotlinx.coroutines.Flow] is created. Flows are lazy (cold) so no service calls are 
+                  made until the flow is collected. This also means there is no guarantee that the request is valid until then. Once 
+                  you start collecting the flow, the SDK will lazily load response pages by making service calls until there are no 
+                  pages left or the flow is cancelled. If there are errors in your request, you will see the failures only after you start
+                  collection.
+                  @param initialRequest A [${inputSymbol.name}] to start pagination
+                  @return a [kotlinx.coroutines.flow.Flow] that can collect [${outputSymbol.name}]
+            """.trimIndent()
+        )
+        writer
+            .addImport(ExternalTypes.KotlinxCoroutines.Flow)
+            .addImport(ExternalTypes.KotlinxCoroutines.FlowGenerator)
+            .addImport(serviceSymbol)
+            .addImport(inputSymbol)
+            .addImport(outputSymbol)
+            .addImport(cursorSymbol)
+            .addImportReferences(cursorSymbol, SymbolReference.ContextOption.DECLARE)
+            .withBlock(
+                "fun #T.#LPaginated(initialRequest: #T): Flow<#T> =",
+                "",
+                serviceSymbol,
+                operationShape.defaultName(),
+                inputSymbol,
+                outputSymbol
+            ) {
+                withBlock("flow {", "}") {
+                    write("var cursor: #F = null", cursorSymbol)
+                    write("var isFirstPage: Boolean = true")
+                    write("")
+                    withBlock("while (isFirstPage || (cursor?.isNotEmpty() == true)) {", "}") {
```

Comment:
**Question**: The cursor symbol is detected above but `isNotEmpty()` seems like it can only apply to specific types. Will this always be a string? If so, do we need to go through the trouble of keeping the cursor type dynamic?

Reply:
> Will this always be a string?
Yes, according to the Smithy spec on the `paginated` trait.

> If so, do we need to go through the trouble of keeping the cursor type dynamic?
This is simply using the model to generate it's specified type.  What do you find troublesome, the action to retrieve the type of the member, using a variable to store that, or the notation in `CodeWriter`?

Reply:
All of the above. In cases where we know the one-and-only type of a variable, we typically hardcode that (e.g., `var isFirstPage: Boolean`). The resulting codegen code is much more readable.

Reply:
is cursor the next token? This is not always a string. See:
```
        "com.amazonaws.dynamodb#Key": {
            "type": "map",
            "key": {
                "target": "com.amazonaws.dynamodb#AttributeName"
            },
            "value": {
                "target": "com.amazonaws.dynamodb#AttributeValue"
            }
        },
```
which is used as a token for DynamoDb::scan